### ISSUE TITLE
Hogs resources when filing out a GitHub issue

### ISSUE THREAD

**Issue by Minoru** [cmcaine/tridactyl#311]
Tridactyl 1.7.3pre665, Firefox 58.0.1 on NixOS

Steps to reproduce:

1. Click "New issue" in any GitHub repository.
2. Start typing in the subject line or body textbox.

Expected behaviour: text appears immediately, Firefox' CPU and memory usage stay roughly constant.

Actual behaviour: after about 10 letters (roughly 2 seconds) the input starts to lag, CPU usage spikes (on multiple cores!), Firefox starts to hog memory.

Workaround: after typing a few words, stop and let Firefox process them. In my case, it takes about 5 seconds for it to get back to its previous CPU and memory usage. Then I can type a few more words.

I disabled all other add-ons and the issue still occurs. I can't reproduce it if I disable Tridactyl.

I'm not entirely sure, but I don't seem to remember seeing this issue at the beginning of January.

**Comment by bovine3dom** [comment]
I can reproduce this on Windows, sort of - CPU usage spikes while typing in any input box with Tridactyl enabled. I don't experience any perceptible lag, though.

**Comment by lazysoundsystem** [comment]
I've found this with Trello too, using Ubuntu. It doesn't happen always, but sometimes severely enough to force a system restart.

**Comment by cmcaine** [comment]
This is interesting. Someone should set breakpoints in the keydown_content
and keydown_background scripts and debug this.

Expected behaviour is that each keypress triggers one trip from
keydown_content, through messaging, keydown_background, controller and
input mode parser.

If more than one trip is generated somewhere or if some of that logic loops
somehow, that could be the lag.

I don't recall anything expensive on that path, but hopefully debugging or
profiling will help us understand the issue.



On 16 Feb 2018 09:27, "Andy Footner" <notifications@github.com> wrote:

I've found this with Trello too, using Ubuntu. It doesn't happen always,
but sometimes severely enough to force a system restart.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/cmcaine/tridactyl/issues/311#issuecomment-366184594>,
or mute the thread

**Comment by pjf** [comment]
I get this under 58.0.2 as well. I'm watching FF chew many more gigabytes whenever I type into this text-field right now.

Having said that, I haven't isolated this to tridactyl, it could be some weird interaction (or another plugin entirely), but the symptoms are the same, and it makes typing all that more exciting from the risk of out-of-memory events!

**Comment by cmcaine** [comment]
I think I'm getting this now, but it locks my laptop up so much it's pretty
hard to debug. I'm going to go through the relevant code carefully this
week.

**Comment by cmcaine** [comment]
I'm definitely getting it now, but can't reliably reproduce.

Can someone who can reliably repro do a git bisect?

Minoru thought it had something to do with having many tabs open. @glacambre recently git bisected for another issue. Feel free to ask me or them for help on how to do this.

**Comment by glacambre** [comment]
I can see CPU usage being higher than without Tridactyl but this happens even on 1.0.0 for me. I can't see anything particularly wrong with the code, I even tried to replace the `ParserController` generator in `controller.ts` with a simple stateful object to see if the issue was just with Firefox not being able to optimize generatos but I had no luck.

A quick and dirty fix is sticking
```javascript
if (state.mode === 'insert' && ke.key !== 'Escape')
    return
```
in `keydown_content.ts`'s `keyeventHandler()`, just after the isTrusted check. There still is a bit of overhead compared to a Tridactylless Firefox, but I think it won't be noticeable even on the slowest of computers.

**Comment by cmcaine** [comment]
Bisected down to the state.ts code, as hypothesised to Minoru last week.

Will push a temporary fix along the lines suggested by @glacambre above. Better fix will come later this week.

For future reference:

Annotated bisect log:

```bash
# bad: [40434455ec8473f1c080072d93abe03bcc1ace7a] release 1.8.0
git bisect start 'HEAD'
# good: [5964430fe10b6db1c35f9b8b3219c8715e7846e6] Convert to es6 modules
git bisect good 5964430fe10b6db1c35f9b8b3219c8715e7846e6
# bad: [8d771e578c3f4cb38bfe26c557ced48c67c66d85] completions 1
git bisect bad 8d771e578c3f4cb38bfe26c557ced48c67c66d85
# bad: [8d771e578c3f4cb38bfe26c557ced48c67c66d85] completions 1
git bisect bad 8d771e578c3f4cb38bfe26c557ced48c67c66d85
# skip: [b2ee48ef0027a0c17b9a432b16686cdbcde0cfd5] Merge branch 'master' of github.com:cmcaine/tridactyl
git bisect skip b2ee48ef0027a0c17b9a432b16686cdbcde0cfd5
# skip: [b2ee48ef0027a0c17b9a432b16686cdbcde0cfd5] Merge branch 'master' of github.com:cmcaine/tridactyl
git bisect skip b2ee48ef0027a0c17b9a432b16686cdbcde0cfd5
# good: [f5cd9d4d7856e6ca9f62a97c28f580a89ee7d2de] Since no scrollx, horizontal scroll with scrollpx
git bisect good f5cd9d4d7856e6ca9f62a97c28f580a89ee7d2de
# good: [f5cd9d4d7856e6ca9f62a97c28f580a89ee7d2de] Since no scrollx, horizontal scroll with scrollpx
git bisect good f5cd9d4d7856e6ca9f62a97c28f580a89ee7d2de
# good: [e01d34a329cc999dbf2a9a720628f1e135a5b0c9] Remove old example CSS, hide URL pop-ups
git bisect good e01d34a329cc999dbf2a9a720628f1e135a5b0c9
# bad: [75ede02bc6ec73605c0fc6c2b0125643e1807a8c] keydown: improve suppression logic
git bisect bad 75ede02bc6ec73605c0fc6c2b0125643e1807a8c
# bad: [504f5539262ebbf7c9da53ce3804ae1bfa49beb7] build: Emit useless & in webpack config
git bisect bad 504f5539262ebbf7c9da53ce3804ae1bfa49beb7

# bad: [2b8d3f0b9d4596e3a3ea81f65d9ebd1749e957c3] AMO: Trim manifest to make reviewing easier
# This is less bad than the above. 50ish load rather than 80 ish. There are 18 commits between these two.
git bisect bad 2b8d3f0b9d4596e3a3ea81f65d9ebd1749e957c3

# bad: [7697072be1ce9ebf6b265ae272806e0a0650bdab] build: Move scripts out. Make generated files obv.
# Was a false positive.
git bisect good 7697072be1ce9ebf6b265ae272806e0a0650bdab
# good: [6fbf5fb9969a9b71729e971df15d82fc9d038046] Merge branch 'master' of github.com:cmcaine/tridactyl
git bisect good 6fbf5fb9969a9b71729e971df15d82fc9d038046
# bad: [927e19c818ab299aaed6ea547611c8cd62dc5a76] hinting: First working version
# Another 45.
git bisect bad 927e19c818ab299aaed6ea547611c8cd62dc5a76
# Skipped commits that didn't compile
# skip: [2c94547fed80236cd6b05fee94bc004c3efcb4e1] src: add initial keyseq.ts
git bisect skip 2c94547fed80236cd6b05fee94bc004c3efcb4e1
# skip: [fad94a697e51b16bbf0843b74dfc093c4365e4b1] parsers: split into real modules
git bisect skip fad94a697e51b16bbf0843b74dfc093c4365e4b1
# skip: [adfca7cf69dfedb8511be045ce0f7a0353c33154] hinting WIP: plumb to background
git bisect skip adfca7cf69dfedb8511be045ce0f7a0353c33154
# good: [b0b809b2d528f325307966309e7994ace3d64c1a] docs: copy docs to build dir!
# This one didn't compile either until I fixed the ModeType typo
git bisect good b0b809b2d528f325307966309e7994ace3d64c1a
# first bad commit: [927e19c818ab299aaed6ea547611c8cd62dc5a76] hinting: First working version
# This also adds state.ts
```

Tested with `web-ext run -s build` with 50 tabs open all on github.com. if `vmstat -Sm1` reported higher than 20 userspace cpu load while I mashed the keyboard I marked the commit bad.

On each commit change I ran `npm run clean && npm install && npm run build`

There are some bugs with this workflow:

1. web-ext won't often automatically update the addon, so you need to go into that window and type R
2. web-ext sometimes loads two copies of the addon. I left a window open on the installed extensions tab and quit the whole bisect process if another one turned up
3. web-ext sometimes becomes unresponsive (pressing r doesn't update the "last extension reload" timestamp). In this case I had to quit firefox and try again

`git bisect log` and `git bisect replay` allow you to recover from web-ext being shit and you marking the wrong commits as good or bad.

**Comment by cmcaine** [comment]
Glacambre's fix didn't seem to work (actually just web-ext fucking up again), so I looked into the code a bit more and fixed it. It's exactly the issue I described to Minoru a week ago - if I'd just looked at the code first I'd have fixed this hours ago. At least I learnt how to use git bisect on a real problem.

Commit message:

Fix #311 - Don't write to browser.storage so often

Turns out writing to browser.storage is expensive because state.ts and
config.ts both listen for it in every tab. I'd vaguely assumed that
Firefox would run other tabs as low priority tasks and that this would
not be an issue, but it doesn't.

I also wrote the code in a rush, and while I mused that this potential
risk exists I didn't document it or explore it.

Retrospective: Content switches are expensive - think before you do
anything that may wake many processes and don't make wild assumptions
about how Firefox handles multiprocessing.

---

Retrospective should have also included that I probably shouldn't write code in a massive rush, and if I do that it should be reviewed, that maybe we should make sure intermediate commits in master are compilable (mostly my fault).
